xio-core/src/main/java/com/xjeffrose/xio/http/Http2MessageSession.java

```diff
@@ -0,0 +1,113 @@
+package com.xjeffrose.xio.http;
+
+import com.google.common.collect.Maps;
+import com.xjeffrose.xio.http.internal.MessageMetaState;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.util.AttributeKey;
+import java.util.Map;
+import javax.annotation.Nullable;
+import lombok.extern.slf4j.Slf4j;
+
+/**
+ * A finite state machine to track the current HTTP/2 message session. This class exists to store
+ * the connection specific state of the message session
+ */
+@Slf4j
+public class Http2MessageSession {
+
+  private static final AttributeKey<Http2MessageSession> CHANNEL_MESSAGE_SESSION_KEY =
+      AttributeKey.newInstance("xio_channel_h2_message_session");
+
+  static Http2MessageSession contextMessageSession(ChannelHandlerContext ctx) {
+    Http2MessageSession session = ctx.channel().attr(CHANNEL_MESSAGE_SESSION_KEY).get();
+    if (session == null) {
+      session = new Http2MessageSession();
+      ctx.channel().attr(CHANNEL_MESSAGE_SESSION_KEY).set(session);
+    }
+    return session;
+  }
+
+  private Map<Integer, MessageMetaState> streamIdRequests = Maps.newHashMap();
+
+  private Http2MessageSession() {}
+
+  public void onRequest(Request request) {
+    MessageMetaState initialRequest = streamIdRequests.get(request.streamId());
+    if (initialRequest == null) {
+      if (request.startOfMessage()) {
+        streamIdRequests.put(
+            request.streamId(), new MessageMetaState(request, request.endOfMessage()));
+      } else {
+        log.error("Received an h2 message segment without a startOfMessage - request: {}", request);
```

Comment:
This wording is a little confusing, maybe something more like "Received h2 Request object without startOfMessage == true".